xarray/backends/zarr.py

```diff
@@ -0,0 +1,535 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from itertools import product
+from base64 import b64encode
+
+import numpy as np
+
+from .. import Variable
+from ..core import indexing
+from ..core.utils import FrozenOrderedDict, HiddenKeyDict
+from ..core.pycompat import iteritems, OrderedDict, integer_types
+from .common import (AbstractWritableDataStore, BackendArray)
+from .. import conventions
+
+# need some special secret attributes to tell us the dimensions
+_DIMENSION_KEY = '_ARRAY_DIMENSIONS'
+
+
+# zarr attributes have to be serializable as json
+# many xarray datasets / variables have numpy arrays and values
+# these functions handle encoding / decoding of such items
+def _encode_zarr_attr_value(value):
+    if isinstance(value, np.ndarray):
+        encoded = value.tolist()
+    # this checks if it's a scalar number
+    elif isinstance(value, np.generic):
+        encoded = value.item()
+        # np.string_('X').item() returns a type `bytes`
+        # zarr still doesn't like that
+        if type(encoded) is bytes:
+            encoded = b64encode(encoded)
+    else:
+        encoded = value
+    return encoded
+
+
+def _ensure_valid_fill_value(value, dtype):
+    if dtype.type == np.string_ and type(value) == bytes:
+        valid = b64encode(value)
+    else:
+        valid = value
+    return _encode_zarr_attr_value(valid)
+
+
+def _decode_zarr_attr_value(value):
+    return value
+
+
+def _decode_zarr_attrs(attrs):
+    return OrderedDict([(k, _decode_zarr_attr_value(v))
+                        for k, v in attrs.items()])
+
+
+def _replace_slices_with_arrays(key, shape):
+    """Replace slice objects in vindex with equivalent ndarray objects."""
+    num_slices = sum(1 for k in key if isinstance(k, slice))
+    array_subspace_size = max(
+        (k.ndim for k in key if isinstance(k, np.ndarray)), default=0)
```

Comment:
@shoyer: Evidently the `default` keyword in `max` is not allowed in python 2.7.

Reply:
I have tried several solutions, but I am embarrassingly stuck on the problem of how to make this `max` call 2/3 compatible. I wish I could just do `from future import max`. 

Help needed.

Reply:
@shoyer do you have any thoughts on how to fix this